### DISCUSSION TITLE
Spacy error in loading pretrained custom model with entity rulers and ner pipeline

### DISCUSSION THREAD

**Discussion by smisra1** [explosion/spaCy#8350] [Help: Coding & Implementations] [answered]
I am using spacy version 2.2.4. I used a spacy blank model with Gensim custom word vectors. Then I trained the model to get the pipeline in the respective order-

entity ruler1, ner1, entity ruler2, ner2

After training it, I saved it in a folder through-

nlp.to_disk('path to folder')

However, if I try to load the same model using nlp1 = spacy.load('path to folder') It gives me this error-

ValueError: [E109] Model for component 'ner' not initialized. Did you forget to load a model, or forget to call begin_training()?

But if I disable both the entity rulers and save ONLY the ner pipes, then it loads successfully.
I cannot find any solution online. What might be the reason I am getting this? How do I successfully load and use my pretrained model which includes entity ruler AND ner both?

**Comment by svlandeg** [comment, marked as answer]
Hi!

Would you be able to provide a small snippet of code that runs by itself and exhibits the error? This will help us debug the issue. Right now there's not really enough information, and I'm confused by the fact that the cited error message refer to the component `"ner"` while you mention having the names `"ner1"` and `"ner2"` in the pipeline. A reproducible code snippet will help clear that up.

Also, please [markdown formatting](https://www.markdownguide.org/basic-syntax/) for code etc, as that makes your post more readable, thanks!

**Reply by smisra1** [reply]
Hi, sorry for the late reply.
Sending the code that runs by itself is a bit difficult. 
The problem is that I am trying to load a pretrained nlp model from my google drive on google colab. This nlp model contains pipes of entity ruler and ner pipes, having custom names.
I can save the pipeline successfully, but I can't load it anywhere. I keep getting that error if I load the nlp model that contains the entity rulers in the pipeline.I have attached a screenshot of the error message. What can I do?
![nlp ent error](https://user-images.githubusercontent.com/83804524/122709984-f863cd80-d27c-11eb-9e03-62bfb5ec1bc8.JPG)

**Reply by polm** [reply]
Please don't post screenshots of code or errors, just copy/paste it as text.

**Reply by svlandeg** [reply]
I can't see the full error stack, so this is just a guess: is the error being thrown from the EntityRuler's `add_patterns` function when it's being deserialized?

What happens in that function is that the phrase patterns are run through `nlp.pipe`, but we're disabling the components further down in the pipeline, because they might not be initialized yet. But there used to be a bug around that if your EntityRuler wasn't actually called `entity_ruler`, which seems to be the case here.

It's this bug & fix: https://github.com/explosion/spaCy/pull/6524

Would you be able to upgrade to spaCy 2.3.5, 2.3.6 of 2.3.7? I think that'll fix this.

**Reply by smisra1** [reply]
Sure, I'll upgrade to a newer version and check. Thank you for your help!

**Reply by smisra1** [reply]
> Please don't post screenshots of code or errors, just copy/paste it as text.

Noted. Sorry, I'm new to this.

**Reply by smisra1** [reply]
> I can't see the full error stack, so this is just a guess: is the error being thrown from the EntityRuler's `add_patterns` function when it's being deserialized?
> 
> What happens in that function is that the phrase patterns are run through `nlp.pipe`, but we're disabling the components further down in the pipeline, because they might not be initialized yet. But there used to be a bug around that if your EntityRuler wasn't actually called `entity_ruler`, which seems to be the case here.
> 
> It's this bug & fix: #6524
> 
> Would you be able to upgrade to spaCy 2.3.5, 2.3.6 of 2.3.7? I think that'll fix this.

Hi. I updated to spacy version 2.3.7 and it worked! Thank you for your help.

**Reply by svlandeg** [reply]
Happy to hear it, thanks for reporting back!
